staging/src/k8s.io/kubectl/pkg/cmd/cp/cp.go

```diff
@@ -349,34 +349,66 @@ func (o *CopyOptions) copyFromPod(src, dest fileSpec) error {
 		return errFileCannotBeEmpty
 	}
 
-	reader, outStream := io.Pipe()
+	reader := newTarPipe(src, o)
+	prefix := getPrefix(src.File)
+	prefix = path.Clean(prefix)
+	// remove extraneous path shortcuts - these could occur if a path contained extra "../"
+	// and attempted to navigate beyond "/" in a remote filesystem
+	prefix = stripPathShortcuts(prefix)
+	return o.untarAll(src, reader, dest.File, prefix)
+}
+
+type TarPipe struct {
+	src       fileSpec
+	o         *CopyOptions
+	reader    *io.PipeReader
+	outStream *io.PipeWriter
+	bytesRead uint64
+}
+
+func newTarPipe(src fileSpec, o *CopyOptions) *TarPipe {
+	t := new(TarPipe)
+	t.src = src
+	t.o = o
+	t.initReadFrom(0)
+	return t
+}
+
+func (t *TarPipe) initReadFrom(n uint64) {
+	t.reader, t.outStream = io.Pipe()
 	options := &exec.ExecOptions{
 		StreamOptions: exec.StreamOptions{
 			IOStreams: genericclioptions.IOStreams{
 				In:     nil,
-				Out:    outStream,
-				ErrOut: o.Out,
+				Out:    t.outStream,
+				ErrOut: t.o.Out,
 			},
 
-			Namespace: src.PodNamespace,
-			PodName:   src.PodName,
+			Namespace: t.src.PodNamespace,
+			PodName:   t.src.PodName,
 		},
 
 		// TODO: Improve error messages by first testing if 'tar' is present in the container?
-		Command:  []string{"tar", "cf", "-", src.File},
+		Command:  []string{"sh", "-c", fmt.Sprintf("tar cf - %s | tail -c+%d", t.src.File, n)},
```

Comment:
I don't know how related is to this, but Ben and I had to spend a lot of time to make it work flawlessly, to stream a tar over a network pipe, check if you can find something useful for this https://github.com/kubernetes-sigs/kind/blob/b6bc112522651d98c81823df56b7afa511459a3b/pkg/cluster/internal/logs/logs.go#L39-L60

Reply:
Yeah nice trick the shell invocation to have access to pipes...
I'm using `tail` to skip the first n bytes of the archive when resuming, however the really nice trick if the pattern where I use a reader of a reader to catch errors and keep track of the transferred data to reconnect without `untalAll` noticing.

Reply:
that is only valid if the source has not changed during the re connection, right?

Reply:
yup, in this respect I don't improve on current implementation - however if your use case is to retrieve GBs of backups, usability is much higher
I might add a flag to set a max retry, but for the moment I'm waiting for some feedback on the issue with a prebuilt binary

Reply:
For backwards compatibility and when not using retries (the proposed default, above) please use only tar, since tail isn't required in that case.